Look_at reimplementation

Goober5000 opened the pull request:
NOTE: This isn't quite ready to be merged yet!  See comments below.

This PR re-implements the look_at feature, which originally was a brute-force iteration down the submodel hierarchy, just like dumb_rotation.  The look_at code now properly uses the `model_instance` code developed by Swifty to integrate into the rest of the collision and rendering code.

A look-at rotation is a type of intrinsic_rotation, just like dumb-rotations. The same code handles both; the only difference is whether submodel_rotate or submodel_look_at is called to apply the actual movement.

The previous look_at parsing code has been optimized by performing the name-to-submodel resolution when the model is actually parsed, instead of when the first look-at rotation is called.  This avoids the need to store a name in the submodel struct and also avoids cluttering the look_at function.

A few functions - `model_instance_find_obj_dir`, `find_submodel_instance_point`, and `find_submodel_instance_point_orient` - have been tweaked to allow them to be called for any model with a model_instance, not just ship models.

This PR is essentially bookkeeping, except for the submodel_look_at method.  Bobboau's previous code has a few flaws - it doesn't work with the new model_instance code, it only accounts for rotation in one level of the submodel hierarchy, it makes an unnecessary detour out to world coordinates - so it can't be used as-is.  Unfortunately, it's also almost entirely inscrutable, with one- or two-letter variables and no comments. :-/  So it was necessary to rewrite the rotation calculation code.

Currently, the rotation code has a few bugs that need to be worked out before committing (assistance with the finer points of vector math is appreciated).  The rotation has to work properly for all three axes, and it also has to work for all ranges of angles.  Currently, only the X axis can be tested with the Praetor model, and currently the door struts flip 180 degrees in the process of opening.

The praetor test mod has been updated since the dumb-rotation PR.  It can be downloaded here:
http://scp.indiegames.us/temp/praetor.zip

asarium commented:
If possible, renaming dumb rotations to Intrinsic rotations should be done in a separate PR. Mixing the look_at reimplementation and the renaming makes code review very hard.

asarium commented on code/model/modelread.cpp:
```diff
@@ -2290,7 +2306,41 @@ int read_model_file(polymodel * pm, char *filename, int n_subsystems, model_subs
+		}
+
+		// done with temporary name vector
+		look_at_submodel_names.clear();
```
Clearing the vector isn't necessary because it will be deallocated anyway when the function finishes.

asarium commented on code/model/modelread.cpp:
```diff
@@ -3322,18 +3373,16 @@ void model_find_obj_dir(vec3d *w_vec, vec3d *m_vec, object *pship_obj, int sub_m
-	polymodel_instance *pmi = model_get_instance(Ships[pship_obj->instance].model_instance_num);
-	polymodel *pm = model_get(Ship_info[Ships[pship_obj->instance].ship_info_index].model_num);
+	polymodel_instance *pmi = model_get_instance(model_instance_num);
+	polymodel *pm = model_get(model_num);
```
An Assertion that checks if the model instance actually belongs to the model number could catch some programming errors here.

On seconds thought, `polymodel_instance` contains the model number of the referenced `polymodel` so the `model_num` parameter isn't needed at all.

asarium commented on code/model/modelread.cpp:
```diff
@@ -3395,24 +3451,26 @@ void model_rot_sub_into_obj(vec3d * outpnt, vec3d *mpnt,polymodel *pm, int sub_m
 		vm_vec_make(model_axis, 0.0f, 0.0f, 1.0f);
+	} else {
+		// must be one of these axes
+		Int3();
```
This should be replaced by an Assertion that includes a better description of the problem. `Int3()` simply crashes the process without any information about the cause.

asarium commented on code/model/modelread.cpp:
```diff
@@ -3521,60 +3579,154 @@ void submodel_stepped_rotate(model_subsystem *psub, submodel_instance_info *sii)
 }
 
-void world_find_real_model_point(vec3d *out, vec3d *world_pt, polymodel *pm, int submodel_num, matrix *orient, vec3d *pos);
+void ez_debug(int index, vec3d *v1, vec3d *v2)
```
This looks like debugging code so I'm going to assume that it will be removed for the final version of the code.

asarium commented on code/model/modelread.cpp:
```diff
@@ -4203,11 +4350,16 @@ void find_submodel_instance_point(vec3d *outpnt, object *pship_obj, int submodel
-	polymodel_instance *pmi = model_get_instance(Ships[pship_obj->instance].model_instance_num);
-	polymodel *pm = model_get(Ship_info[Ships[pship_obj->instance].ship_info_index].model_num);
+	polymodel_instance *pmi = model_get_instance(model_instance_num);
+	polymodel *pm = model_get(model_num);
```
Same as above, either `model_num` should be validated or the parameter should be removed.

asarium commented on code/model/modelread.cpp:
```diff
@@ -4319,12 +4471,17 @@ void find_submodel_instance_point_orient(vec3d *outpnt, matrix *outorient, objec
-	polymodel_instance *pmi = model_get_instance(Ships[pship_obj->instance].model_instance_num);
-	polymodel *pm = model_get(Ship_info[Ships[pship_obj->instance].ship_info_index].model_num);
+	polymodel_instance *pmi = model_get_instance(model_instance_num);
+	polymodel *pm = model_get(model_num);
```
See above.

asarium commented on code/model/modelread.cpp:
```diff
@@ -4790,52 +4947,55 @@ void model_update_instance(int model_instance_num, int sub_model_num, submodel_i
 		polymodel *pm = model_get(pmi->model_num);
-		bsp_info *sm = &pm->submodel[sub_it->submodel_num];
+		bsp_info *sm = &pm->submodel[submodel_it->submodel_num];
 		Assert(pm != nullptr && sm != nullptr);
```
Checking `sm != nullptr` isn't really helpful here because in pretty much any case it will be `!= nullptr`.

asarium commented:
I added some comments but I haven't looked at the vector math yet. I'll do that at home when I can see the whole code in a real editor.

Goober5000 commented:
PR #531 created, which is basically just the refactor of dumb_rotation to intrinsic_rotation.  Once that PR is approved, I'll rebase this one.

Goober5000 commented:
PR rebased, and I added a new commit which addressed most of the comments here.  I'm reluctant to merge the `model_num` and `model_instance_num` parameters because there are about twelve function prototypes scattered throughout the code that use both.

I still need to fix the angle calculations, but I've heard back from Bobboau and I have another math guy willing to look at it, so the investigation is progressing.

asarium commented:
I understand the concerns. Adding `Assertion(model_num == pmi->model_num, "Instance model number does not match the specified model number!");` would catch any coding mistakes without changing the function signature.

Goober5000 commented:
Again, if you're going to do that once, you might as well do it in all twelve functions.  That's better suited to another PR.

I suppose I could make such a PR.

asarium commented:
That's even better! My suggestion was to just validate the instance and model number without changing the function signature but removing the parameter is the better solution.

Goober5000 commented:
Yeah, I figured, if you want a PR anyway, might as well use the better solution.

The model instance code is fairly well debugged by now, but there is admittedly a possibility that the parameters could be transposed.  This mitigates that risk.

Looks like there were some problems.  I'll fix those.  Once that PR is merged, I'll rebase this one again.

Goober5000 commented:
I just figured out that the reason I'm having so much trouble with the vector math could be that the *test model itself* is set up incorrectly. :-/

In addition to preventing my current tests from working, this would have required Bobboau's original code to not only be inscrutable, but also use incorrect calculations.

I shall be in contact with a modeler shortly to deliver a test model according to my personal specifications.

Goober5000 commented:
I've posted a model request:
http://www.hard-light.net/forums/index.php?topic=91507.0

Goober5000 commented:
Closing in favor of #3144.
